shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1696,116 +1675,121 @@ - (void)startKeyBoardAnimation:(NSTimeInterval)duration {
+        if (!strongSelf.keyboardAnimationView.superview) {
+          // Ensure the keyboardAnimationView is in view hierarchy when animation running.
+          [strongSelf.view addSubview:strongSelf.keyboardAnimationView];
+        }
 
-    if ([flutterViewController keyboardSpringAnimation] == nil) {
-      if (flutterViewController.get().keyboardAnimationView.layer.presentationLayer) {
-        flutterViewController.get()->_viewportMetrics.physical_view_inset_bottom =
-            flutterViewController.get()
-                .keyboardAnimationView.layer.presentationLayer.frame.origin.y;
-        [flutterViewController updateViewportMetricsIfNeeded];
-      }
-    } else {
-      fml::TimeDelta timeElapsed =
-          keyboardAnimationTargetTime - flutterViewController.get().keyboardAnimationStartTime;
-      flutterViewController.get()->_viewportMetrics.physical_view_inset_bottom =
-          [[flutterViewController keyboardSpringAnimation] curveFunction:timeElapsed.ToSecondsF()];
-      [flutterViewController updateViewportMetricsIfNeeded];
-    }
-  };
+        if (!strongSelf.keyboardSpringAnimation) {
+          if (strongSelf.keyboardAnimationView.layer.presentationLayer) {
+            strongSelf->_viewportMetrics.physical_view_inset_bottom =
+                strongSelf.keyboardAnimationView.layer.presentationLayer.frame.origin.y;
+            [strongSelf updateViewportMetricsIfNeeded];
+          }
+        } else {
+          fml::TimeDelta timeElapsed =
+              keyboardAnimationTargetTime - strongSelf.keyboardAnimationStartTime;
+          strongSelf->_viewportMetrics.physical_view_inset_bottom =
+              [strongSelf.keyboardSpringAnimation curveFunction:timeElapsed.ToSecondsF()];
+          [strongSelf updateViewportMetricsIfNeeded];
+        }
+      };
   [self setUpKeyboardAnimationVsyncClient:keyboardAnimationCallback];
   VSyncClient* currentVsyncClient = _keyboardAnimationVSyncClient;
 
   [UIView animateWithDuration:duration
       animations:^{
+        FlutterViewController* strongSelf = weakSelf;
+        if (!strongSelf) {
+          return;
+        }
+
         // Set end value.
-        [self keyboardAnimationView].frame = CGRectMake(0, self.targetViewInsetBottom, 0, 0);
+        strongSelf.keyboardAnimationView.frame = CGRectMake(0, self.targetViewInsetBottom, 0, 0);
 
         // Setup keyboard animation interpolation.
         CAAnimation* keyboardAnimation =
-            [[self keyboardAnimationView].layer animationForKey:@"position"];
-        [self setUpKeyboardSpringAnimationIfNeeded:keyboardAnimation];
+            [strongSelf.keyboardAnimationView.layer animationForKey:@"position"];
+        [strongSelf setUpKeyboardSpringAnimationIfNeeded:keyboardAnimation];
       }
       completion:^(BOOL finished) {
         if (_keyboardAnimationVSyncClient == currentVsyncClient) {
+          FlutterViewController* strongSelf = weakSelf;
+          if (!strongSelf) {
+            return;
+          }
```

Comment:
In theory we can just call the three methods on weakSelf here, but I'm assuming if we call one we want to call all three and didn't want to think about threading.

Reply:
Xcode has a special warning just for this; the way you wrote it is definitely better.